tests/integration_tests/test_public_api.py

```diff
@@ -0,0 +1,145 @@
+import ipaddress
+from contextlib import contextmanager, suppress
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Generator
+
+import pytest
+
+import pycloudlib
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.instance import BaseInstance
+from pycloudlib.util import LTS_RELEASES
+
+cloud_config = """\
+#cloud-config
+runcmd:
+  - echo 'hello' >> /var/tmp/example.txt
+"""
+
+
+@pytest.fixture
+def cloud(request):
+    cloud_instance: BaseCloud = request.param(
+        tag="pycl-test",
+        timestamp_suffix=True,
+    )
+
+    if isinstance(cloud_instance, pycloudlib.EC2):
+        cloud_instance.upload_key(
+            public_key_path=cloud_instance.config["public_key_path"],
+            private_key_path=cloud_instance.config["private_key_path"],
+            name=cloud_instance.tag,
+        )
+
+    yield cloud_instance
+
+    if isinstance(cloud_instance, pycloudlib.EC2):
+        cloud_instance.delete_key(name=cloud_instance.tag)
+
+
+@contextmanager
+def launch_instance(
+    cloud_instance: BaseCloud, **kwargs
+) -> Generator[BaseInstance, None, None]:
+    instance = cloud_instance.launch(**kwargs)
+    try:
+        yield instance
+    finally:
+        instance.delete()
```

Comment:
Thoughts on observing a `PYCLOUDLIB_KEEP_INSTANCE=0` env var or `pycloudlib.toml` config param to avoid teardown after test completion to aid in development of integration tests? Or do we just think it's best for folks to drop a pdb in to prevent this sort of teardown during development.

```suggestion
        if bool(os.environ.get("PYCLOUDLIB_KEEP_INSTANCE", False)):
             instance.delete()
        else:
            LOG.info(
                f"Leaving instance up at {instance.name}: ({instance.ip})"
            )
```

Reply:
At this point, I think I'd prefer we use pdb. I think we'd need a bit more test infrastructure for something like that, otherwise every integration test will be implementing this sort of thing on its own.

I think it made a lot more sense as an option in cloud-init because we were testing what happens between launch and teardown. Since here we're testing the launch and teardown themselves (along with other instance lifecycle), I'm not sure we'll need as much "in the middle" debugging. I could be wrong, and am certainly open to adding that if we need it, but for now I think it's better to keep it simple.